server/api/diagnose.go

```diff
@@ -0,0 +1,156 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/pd/server"
+	"github.com/unrolled/render"
+	"net/http"
+	"time"
+)
+
+type diagnoseType int
+type diagnoseKV map[diagnoseType]Recommended
+
+//Recommended return detail
+type Recommended struct {
+	Module      string `json:"module"`
+	Level       string `json:"level"`
+	Description string `json:"description"`
+	Instruction string `json:"instruction"`
+}
+
+const (
+	//analyze level
+	levelNormal   = "Normal"
+	levelWarning  = "Warning"
+	levelMinor    = "Minor"
+	levelMajor    = "Major"
+	levelCritical = "Critical"
+	levelFatal    = "Fatal"
+
+	//analyze module
+	modMember   = "member"
+	modTiKV     = "TiKV"
+	modReplica  = "Replic"
+	modSchedule = "Schedule"
+	modDefault  = "indefinite"
+
+	memberOneInstance diagnoseType = iota
+	memberEvenInstance
+	memberLostPeers
+	memberLostPeersMoreThanHalf
+	memberLeaderChanged
+	tikvCap70
+	tikvCap80
+	tikvCap90
+	tikvLostPeers
+	tikvLostPeersLongTime
+)
+
+var (
+	diagnoseMap = diagnoseKV{
+		memberOneInstance:           Recommended{modMember, levelWarning, "only one PD instance is running", "please add PD instance"},
+		memberEvenInstance:          Recommended{modMember, levelMinor, "PD instances is even number", "the recommended number of PD's instances is odd"},
+		memberLostPeers:             Recommended{modMember, levelMajor, "some PD instances is down", "please check host load and traffic"},
+		memberLostPeersMoreThanHalf: Recommended{modMember, levelCritical, "more than half PD instances is down", "please check host load and traffic"},
+		memberLeaderChanged:         Recommended{modMember, levelMinor, "PD cluster leader is changed", "please check host load and traffic"},
+		tikvCap70:                   Recommended{modTiKV, levelWarning, "some TiKV stroage used more than 70%", "plase add TiKV node"},
+		tikvCap80:                   Recommended{modTiKV, levelMinor, "some TiKV stroage used more than 80%", "plase add TiKV node"},
+		tikvCap90:                   Recommended{modTiKV, levelMajor, "some TiKV stroage used more than 90%", "plase add TiKV node"},
+		tikvLostPeers:               Recommended{modTiKV, levelWarning, "some TiKV lost connect", "plase check network"},
+		tikvLostPeersLongTime:       Recommended{modTiKV, levelMajor, "some TiKV lost connect more than 1h", "plase check network"},
+	}
+)
+
+type diagnoseHandler struct {
+	svr *server.Server
+	rd  *render.Render
+}
+
+func newDiagnoseHandler(svr *server.Server, rd *render.Render) *diagnoseHandler {
+	return &diagnoseHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+func diagnosePD(key diagnoseType) *Recommended {
```

Comment:
I think we can be sure the key is valid, so the function is unnecessary. It is ok to use `disgnoseMap[key]` directly.

Reply:
this function will add more params,like ` diagnosePD(key diagnoseType, descAdd, instAdd string)`
`Description` and `Description` support additional messages.